### PR TITLE
Assign new material to multiple objects

### PR DESCRIPTION
**Pull request by stefanminning-autodesk**
This PR adds functionality to the existing `UsdUndoAssignNewMaterialCommand`, allowing for a new material to be assigned to multiple objects.

The command's original functionality remains untouched, as I wanted to avoid breaking API changes.

### CODE REVIEW

**Comment by seando-adsk** [comment]
@stefanminning-autodesk Before we start the code review and preflight process can you please clean this up by squashing the 11 commits?

**Comment by stefanminning-autodesk** [comment]
Sure, I've pushed with squashed history.

**Comment by seando-adsk** [comment]
@womanyoyoyo-adsk Natalia can you please have a look at this from a design perspective? 
@stefanminning-autodesk From the code changes it looks to me as if you are always applying the material the selection, regardless of the item the context menu was opened on. Normally the behavior would be:
Q: "is the item on which the context menu was opened in the selection or not"?
Yes - apply operation to selection
No - apply operation only to context item

**Comment by stefanminning-autodesk** [comment]
That's right, I've coded it to match the current behaviour, where a new material would be applied to an object when the right-click operation is applied upon it, regardless of its selection state.

Personally, I also find it somewhat confusing, but I'll leave it to the UX team to decide whether this is a good workflow or not.

Should be an easy change, either way.

**Comment by seando-adsk** [comment]
@stefanminning-autodesk Actually sorry I don't think that is correct. With your changes it applies the material to **all** the selected objects. If the object you right click on is **not** selected, then the material is **not** applied to it. The normal context menu behavior would be "if the context object is selected?" then apply to selection, otherwise apply only to context object.

**Comment by stefanminning-autodesk** [comment]
Ah yes, you're right: Looking at the code again, it should not be ignoring `fItem`. Could have sworn I'd accounted for that at some point... must have slipped through the cracks.

Anyway, guess I'll have to have another look at this when the designers are back from holidays and they've clarified the exact behaviour they'd like to see.

**Comment by womanyoyoyo-adsk** [comment]
Hi, myself and Lookdevx designer are looking into this. Keep you posted.

File: lib/mayaUsd/ufe/UsdUndoMaterialCommands.h
Comment (hodoulp) [inline lib/mayaUsd/ufe/UsdUndoMaterialCommands.h:100]: There is already a predefined type `std::list<std::shared_ptr<SceneItem> > SceneItemList;` that the code can perhaps use.

File: lib/mayaUsd/ufe/UsdUndoMaterialCommands.h
Comment (hodoulp) [inline lib/mayaUsd/ufe/UsdUndoMaterialCommands.h:114]: Not clear about Doxygen but I would have move `\p` close to the parameter name `sdrShaderIdentifier`

File: lib/mayaUsd/ufe/UsdUndoMaterialCommands.cpp
Comment (hodoulp) [inline lib/mayaUsd/ufe/UsdUndoMaterialCommands.cpp:171]: Does that code need `UFE_PREVIEW_XX` defines protection?

File: lib/mayaUsd/ufe/UsdUndoMaterialCommands.h
Comment (hodoulp) [inline lib/mayaUsd/ufe/UsdUndoMaterialCommands.h:118]: Should the new method be protected by `UFE_PREVIEW_XX` defines?

File: lib/mayaUsd/ufe/UsdUndoMaterialCommands.h
Comment (hodoulp) [inline lib/mayaUsd/ufe/UsdUndoMaterialCommands.h:101]: Should the new method be protected by UFE_PREVIEW_XX defines?

**Review by hodoulp — COMMENTED** [review]
LGTM

File: lib/mayaUsd/ufe/UsdUndoMaterialCommands.cpp
Comment (stefanminning-autodesk) [reply lib/mayaUsd/ufe/UsdUndoMaterialCommands.cpp:171]: Good question, honestly not sure. 

These changes are entirely contained in MayaUSD, with no UFE update required. What UFE version would I be looking for?

File: lib/mayaUsd/ufe/UsdUndoMaterialCommands.h
Comment (stefanminning-autodesk) [reply lib/mayaUsd/ufe/UsdUndoMaterialCommands.h:100]: Good one, thank you. Much more readable.

File: lib/mayaUsd/ufe/UsdContextOps.cpp
Comment (seando-adsk) [inline lib/mayaUsd/ufe/UsdContextOps.cpp:1265]: This could be simplified by just using a Ufe Selection (which internally uses a SceneItemList).
```suggestion
        auto sceneItems = Ufe::GlobalSelection::get();
        sceneItems.append(fItem);
```

The Ufe Selection append will only add the item if its not already there. If you really need the item at the front then you can use insert instead. Then just pass this selection to the UsdUndoAssignNewMaterialCommand.

Comment (seando-adsk) [reply lib/mayaUsd/ufe/UsdContextOps.cpp:1265]: No you cannot get the SceneItemList out of it. There are iterators for looping thru. But I would use Ufe::Selection not Ufe::GlobalSelection.

File: lib/mayaUsd/ufe/UsdUndoMaterialCommands.h
Comment (seando-adsk) [inline lib/mayaUsd/ufe/UsdUndoMaterialCommands.h:129]: Sorry I didn't catch this earlier. We shouldn't use the namespace name directly, but instead the define. So "pxr::" should be "PXR_NS::" (see line 126).

Comment (seando-adsk) [reply lib/mayaUsd/ufe/UsdUndoMaterialCommands.h:129]: Once this and the merge conflict is fixed, you can run another preflight and if it passes add the label "ready for merge" and I'll merge it. I confirmed with our designer that the multiple selection workflow is okay.

**Review by seando-adsk — APPROVED** [review]

**Comment by seando-adsk** [comment]
@stefanminning-autodesk There are clang-format errors that need to be fixed. Please see [codingGuidelines](https://github.com/Autodesk/maya-usd/blob/dev/doc/codingGuidelines.md) for info on clang-format.

**Review by seando-adsk — APPROVED** [review]

**Review by seando-adsk — DISMISSED** [review]

File: lib/mayaUsd/ufe/UsdContextOps.cpp
Comment (seando-adsk) [reply lib/mayaUsd/ufe/UsdContextOps.cpp:1279]: Were you by chance getting the Ufe Selection into a reference pointer and them modifying it? You would just need to make sure to get a copy of it (using the copy constructor).

**Review by seando-adsk — CHANGES_REQUESTED** [review]
I think you should also update the test in testContextOps.py to test the multiple selection scenario

File: test/lib/ufe/testContextOps.py
Comment (seando-adsk) [inline test/lib/ufe/testContextOps.py:741]: I think we should have a way to make sure this is automatically restored using a context manager object. Look at `TemporaryDirectory`. You can create one that will restore an env var. The input would be the env var and the new value. The __init__ will get the current value, store it and set the env var to input and the __exit__ will restore to stored value. Then you can wrap this section of code in a `with ...` statement.

**Review by seando-adsk — CHANGES_REQUESTED** [review]
Thanks for the test. If you could just create a new context object and use it the two places that env var is set/restored that would be great. The new context object can be in testUtils.py

**Review by seando-adsk — APPROVED** [review]
Everything looks good now. Thanks.

**Comment by seando-adsk** [comment]
@stefanminning-autodesk FYI, any new commit (such as your latest one merging in dev) will invalidate the preflight check and it must be run again. So I've removed the ready for merge label and re-started the preflight for you. When it passes, you may add back the label and I'll merge.

**Review by seando-adsk — APPROVED** [review]

**Pull request merged**
